### ISSUE TITLE
DbConnection.CloneWith

### ISSUE THREAD

**Issue by rgarrison12345** [dotnet/runtime#39876]
Hello,

Often times I find I have to clone an existing instance of DbConnection object.  But I want to clone it with a different connection string, but still keeping same security information.  Such as password, user name, etc.

public abstract DbConnection CloneWith(string connectionString);

**Comment by KalleOlaviNiemitalo** [comment]
System.Data.SqlClient.SqlConnection already allows a similar feature if you use SqlCredential. The Clone method copies the credential, and changing the ConnectionString of the clone does not clear the credential.

**Comment by rgarrison12345** [comment]
Hello @KalleOlaviNiemitalo, I'm familiar with the SqlClient functionality.  I am proposing adding a function on the abstract class DbConnection. MySqlConnector and NpgSql drivers already have this feature defined with a similar signature I proposed.

MySqlConnector 
https://github.com/mysql-net/MySqlConnector/blob/master/src/MySqlConnector/MySqlConnection.cs

NpgSql 
https://github.com/npgsql/npgsql/blob/dev/src/Npgsql/NpgsqlConnection.cs

**Comment by roji** [comment]
@rgarrison12345 I agree it may make sense to include functionality like this on DbConnection. To set expectations though, we probably won't be able to get this into .NET 5.0 at this point in the release cycle.

**Comment by KalleOlaviNiemitalo** [comment]
Adding an abstract method to `DbConnection` would be a breaking change for existing ADO.NET providers.

The `Clone()` method is not part of `DbConnection`, either. It is part of `ICloneable`, which is implemented by some classes derived from `DbConnection`. A new interface for `CloneWith` might make sense. Alternatively, `DbConnection.CloneWith` could be a virtual method that throws `NotSupportedException` by default, as [recommended in the `NotImplementedException` remarks](<https://docs.microsoft.com/dotnet/api/system.notimplementedexception?view=net-5.0#notimplementedexception-and-other-exception-types> "NotImplementedException and other exception types | NotImplementedException Class (System) | Microsoft Docs").

Do you think applications would want to check in advance whether an ADO.NET provider supports `CloneWith`? It seems to me that, if an application has fallback code for use with providers that do not support `CloneWith`, then it is easier for the application to always use that and not bother with `CloneWith` at all.

**Comment by roji** [comment]
> Adding an abstract method to DbConnection would be a breaking change for existing ADO.NET providers.

As you writ below, we can just add a non-abstract virtual method that throws NotSupportedException.

> Do you think applications would want to check in advance whether an ADO.NET provider supports CloneWith?

We sometimes add capability flags for this, and this may be a good case to do that.

> It seems to me that, if an application has fallback code for use with providers that do not support CloneWith, then it is easier for the application to always use that and not bother with CloneWith at all.

Unless I'm missing something, the imaginable fallback code wouldn't be able to do much beyond copy the connection string from the old to the new connection string (this could even be the default implementation for the new CloneWith method). This would work in some cases (e.g. `Integrated Security=true`, `Persist Security Info=true`), but not in all cases, which is exactly what CloneWith would address (e.g. the default `Persist Security Info=false`). So I do think there's value in CloneWith

**Comment by KalleOlaviNiemitalo** [comment]
> this could even be the default implementation for the new CloneWith method

Wouldn't it be difficult for a default implementation of `DbConnection.CloneWith` to parse parameters from the connection string and know which of them contain security information?

**Comment by roji** [comment]
@KalleOlaviNiemitalo I'm not proposing anything be parsed anywhere (corrected to above which was inaccurate) - only that the default implementation of CloneWith simply instantiate a new connection with the connection string provided by the user. If that connection string itself is sufficient to connect, CloneWith would work - otherwise it would fail. For the cases which are actually interesting, the provider would have to override and implement the behavior.

**Comment by KalleOlaviNiemitalo** [comment]
Such a default implementation would still have to throw NotSupportedException if the derived class does not implement ICloneable.

**Comment by roji** [comment]
Depending on provider, the connection string usually represents many/most of the settable things on a connection, so I do think there's value in a *default* implementation that simply instantiates the connection with the connection string. But that's not a terribly important question, it can also throw instead.
